### DISCUSSION TITLE
Metrics appearing multiple times even if emitted once.

### DISCUSSION THREAD

**Discussion by arijitsurmicrosoft** [open-telemetry/opentelemetry-dotnet#4412] [Q&A] [answered]
_metricCounter.Add(1, new("role", "mtproxy"), new("roleinstance", "azavsuatwebp101"), new("requestType", Constants.GET) , new("resultType", Constants.PIDPROXYPASS), new("durationMs","0"));
 
Hello everyone, the metric is keeping on emitting events even though the method (inside which this statement resides) is called only once.
![imgpsh_mobile_save](https://user-images.githubusercontent.com/100192850/233326758-82520e86-1e95-4ec2-8c56-8a99239fa1e9.jpg)

**Comment by utpilla** [comment, marked as answer]
It looks like you're using ConsoleExporter which by default has a TemporalityPreference as Cumulative and not Delta. With Cumulative aggregation temporality, the MetricReader (ConsoleExporter in this case) would keep exporting the cumulated data for the metric with every collection. Metrics are reported for a certain duration. Note the start time and end time in the output. In case of Cumulative aggregation, the start time doesn't change and only the end time changes.

ConsoleExporter is only meant to provide you a fair idea of how the metrics data get emitted. There are a lot of backend systems that accept Cumulative aggregated data, the most prominent being Prometheus.

**Reply by arijitsurmicrosoft** [reply]
https://opentelemetry.io/docs/instrumentation/python/exporters/#temporality-presets

Thank you for your answer, I was expecting the same. Can you show me a doc which speaks about a way to set the temporalitypreference for exporters. 
I used setx OTEL_EXPORTER_METRICS_TEMPORALITY_PREFERENCE DELTA
but i still see metrics getting emitted multiple times.

**Reply by arijitsurmicrosoft** [reply]
![17cd36f8-d538-4fe5-bea3-fa95f3912d11](https://user-images.githubusercontent.com/100192850/233551367-86b6a2ae-881d-4a61-a399-6f35a1210bc6.jpg)
I have tried both setx and set, but i see the same output.

**Reply by cijothomas** [reply]
https://opentelemetry.io/docs/instrumentation/python/exporters/#temporality-presets --> This is a doc from Python OpenTelemetry. It won't work in .NET apps. (IIRC, the environment variable is not supported in OTel .NET)

**Reply by cijothomas** [reply]
https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.Console/ConsoleExporterMetricsExtensions.cs#L86

.AddConsoleExporter((exporterOptions, metricReaderOptions) =>
                    {
                        metricReaderOptions.TemporalityPreference = MetricReaderTemporalityPreference.Delta;
                    });

^ This shows how to change temporality for Console.

**Comment by cijothomas** [comment]
Once the delta/cumulative is changed, review the usage of metrics dimensions. It looks like one of the dimension is "durationMs", which is a bad choice for dimension. Durations should be the metric value itself, and modelling it as dimension would easily cause data loss due to cardinality capping https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics/customizing-the-sdk#changing-maximum-metricpoints-per-metricstream

**Reply by arijitsurmicrosoft** [reply]
Thank you.
